File: src/models/transaction_item_group.ts
export class TransactionItemGroup implements TransactionItemGroupInfoResponse {
    public id: string;
    public name: string;
    public displayOrder: number;

    private constructor(id: string, name: string, displayOrder: number) {
        this.id = id;
        this.name = name;
        this.displayOrder = displayOrder;
    }

    public toCreateRequest(): TransactionItemGroupCreateRequest {
        return {
            name: this.name
        };
    }

    public toModifyRequest(): TransactionItemGroupModifyRequest {
        return {
            id: this.id,
            name: this.name
        };
    }

    public clone(): TransactionItemGroup {
        return new TransactionItemGroup(this.id, this.name, this.displayOrder);
    }

    public static of(itemGroupResponse: TransactionItemGroupInfoResponse): TransactionItemGroup {
        return new TransactionItemGroup(itemGroupResponse.id, itemGroupResponse.name, itemGroupResponse.displayOrder);
    }

    public static ofMulti(itemGroupResponses: TransactionItemGroupInfoResponse[]): TransactionItemGroup[] {
        const itemGroups: TransactionItemGroup[] = [];

        for (const itemGroupResponse of itemGroupResponses) {
            itemGroups.push(TransactionItemGroup.of(itemGroupResponse));
        }

        return itemGroups;
    }

    public static createNewItemGroup(name?: string): TransactionItemGroup {
        return new TransactionItemGroup('', name || '', 0);
    }
}

export interface TransactionItemGroupCreateRequest {
    readonly name: string;
}

export interface TransactionItemGroupModifyRequest {
    readonly id: string;
    readonly name: string;
}

export interface TransactionItemGroupMoveRequest {
    readonly newDisplayOrders: TransactionItemGroupNewDisplayOrderRequest[];
}

export interface TransactionItemGroupNewDisplayOrderRequest {
    readonly id: string;
    readonly displayOrder: number;
}

export interface TransactionItemGroupDeleteRequest {
    readonly id: string;
}

export interface TransactionItemGroupInfoResponse {
    readonly id: string;
    readonly name: string;
    readonly displayOrder: number;
}

File: pkg/models/transaction_item_group.go
package models

// TransactionItemGroup represents transaction item group data stored in database
type TransactionItemGroup struct {
	ItemGroupId     int64  `xorm:"PK"`
	Uid             int64  `xorm:"INDEX(IDX_item_group_uid_deleted_order) NOT NULL"`
	Deleted         bool   `xorm:"INDEX(IDX_item_group_uid_deleted_order) NOT NULL"`
	Name            string `xorm:"VARCHAR(64) NOT NULL"`
	DisplayOrder    int32  `xorm:"INDEX(IDX_item_group_uid_deleted_order) NOT NULL"`
	CreatedUnixTime int64
	UpdatedUnixTime int64
	DeletedUnixTime int64
}

// TransactionItemGroupGetRequest represents all parameters of transaction item group getting request
type TransactionItemGroupGetRequest struct {
	Id int64 `form:"id,string" binding:"required,min=1"`
}

// TransactionItemGroupCreateRequest represents all parameters of transaction item group creation request
type TransactionItemGroupCreateRequest struct {
	Name string `json:"name" binding:"required,notBlank,max=64"`
}

// TransactionItemGroupModifyRequest represents all parameters of transaction item group modification request
type TransactionItemGroupModifyRequest struct {
	Id   int64  `json:"id,string" binding:"required,min=1"`
	Name string `json:"name" binding:"required,notBlank,max=64"`
}

// TransactionItemGroupMoveRequest represents all parameters of transaction item group moving request
type TransactionItemGroupMoveRequest struct {
	NewDisplayOrders []*TransactionItemGroupNewDisplayOrderRequest `json:"newDisplayOrders" binding:"required,min=1"`
}

// TransactionItemGroupNewDisplayOrderRequest represents a data pair of id and display order
type TransactionItemGroupNewDisplayOrderRequest struct {
	Id           int64 `json:"id,string" binding:"required,min=1"`
	DisplayOrder int32 `json:"displayOrder"`
}

// TransactionItemGroupDeleteRequest represents all parameters of transaction item group deleting request
type TransactionItemGroupDeleteRequest struct {
	Id int64 `json:"id,string" binding:"required,min=1"`
}

// TransactionItemGroupInfoResponse represents a view-object of transaction item group
type TransactionItemGroupInfoResponse struct {
	Id           int64  `json:"id,string"`
	Name         string `json:"name"`
	DisplayOrder int32  `json:"displayOrder"`
}

// ToTransactionItemGroupInfoResponse returns a view-object according to database model
func (t *TransactionItemGroup) ToTransactionItemGroupInfoResponse() *TransactionItemGroupInfoResponse {
	return &TransactionItemGroupInfoResponse{
		Id:           t.ItemGroupId,
		Name:         t.Name,
		DisplayOrder: t.DisplayOrder,
	}
}

// TransactionItemGroupInfoResponseSlice represents the slice data structure of TransactionItemGroupInfoResponse
type TransactionItemGroupInfoResponseSlice []*TransactionItemGroupInfoResponse

// Len returns the count of items
func (s TransactionItemGroupInfoResponseSlice) Len() int {
	return len(s)
}

// Swap swaps two items
func (s TransactionItemGroupInfoResponseSlice) Swap(i, j int) {
	s[i], s[j] = s[j], s[i]
}

// Less reports whether the first item is less than the second one
func (s TransactionItemGroupInfoResponseSlice) Less(i, j int) bool {
	return s[i].DisplayOrder < s[j].DisplayOrder
}
